File: DynamicProgramming/EditDistance/EditDistance.js
function method1(str, tar) {
  let n = str.length,
    m = tar.length,
    i,
    j;
  let mat = new Array(n + 1);
  for (i = 0; i <= n; i++) {
    mat[i] = new Array(m + 1);
    mat[i].fill(0);
  }
  for (i = 0; i <= m; i++) {
    mat[0][i] = i;
  }
  for (i = 0; i <= n; i++) {
    mat[i][0] = i;
  }
  for (i = 1; i <= n; i++) {
    for (j = 1; j <= m; j++) {
      if (str.charAt(i - 1) != tar.charAt(j - 1)) {
        mat[i][j] =
          Math.min(mat[i - 1][j], Math.min(mat[i - 1][j - 1], mat[i][j - 1])) +
          1;
      } else {
        mat[i][j] = mat[i - 1][j - 1];
      }
    }
  }
  //console.log("mat is ", mat);
  return mat[n][m];
}

function method2(str, tar) {
  let i,
    n = str.length,
    m = tar.length,
    j;
  let prev = new Array(m + 1);
  let cur = new Array(m + 1);
  for (i = 0; i <= m; i++) {
    prev[i] = i;
  }
  for (i = 1; i <= n; i++) {
    cur[0] = i;
    for (j = 1; j <= m; j++) {
      if (str.charAt(i - 1) != tar.charAt(j - 1)) {
        cur[j] = Math.min(prev[j],Math.min(prev[j - 1], cur[j - 1]))+1;
      } else {
        cur[j] = prev[j - 1];
      }
    }
    prev = [...cur];
  }
  return prev[m];
}

let str = "geek",
  tar = "gesek";
console.log(method1(str, tar));
console.log(method2(str, tar));

File: DynamicProgramming/EditDistance/EditDistance.go
package main

import (
	"fmt"
)

func min(i int, j int) int {
	if i < j {
		return i
	}
	return j
}

func method1(str string, str2 string) int {
	n := len(str)
	n2 := len(str2)
	arr := make([][]int, n+1)
	for i := 0; i <= n; i++ {
		arr[i] = make([]int, n2+1)
		arr[i][0] = i
	}

	for i := 0; i <= n2; i++ {
		arr[0][i] = i
	}

	for i := 1; i <= n; i++ {
		for j := 1; j <= n2; j++ {
			if str[i-1:i] != str2[j-1:j] {
				arr[i][j] = 1 + min(arr[i-1][j], min(arr[i-1][j-1], arr[i][j-1]))
			} else {
				arr[i][j] = arr[i-1][j-1]
			}
		}
	}
	//fmt.Println("arr is ", arr)
	return arr[n][n2]
}

func method2(str string, str2 string) int {
	n := len(str)
	n2 := len(str2)
	prev := make([]int, n2+1)
	cur := make([]int, n2+1)
	for i := 0; i <= n2; i++ {
		prev[i] = i
	}
	for i := 1; i <= n; i++ {
		cur[0] = i - 1
		for j := 1; j <= n2; j++ {
			if str[i-1:i] != str2[j-1:j] {
				cur[j] = 1 + min(prev[j], min(prev[j-1], cur[j-1]))
			} else {
				cur[j] = prev[j-1]
			}
		}
		copy(prev, cur)
	}
	return cur[n2]
}

func main() {
	str := "sunday"
	tar := "saturday"

	fmt.Println(method1(str, tar))
	fmt.Println(method2(str, tar))
}
